test/python/circuit/test_library.py

```diff
@@ -50,3 +58,449 @@ def test_inner_product(self):
+        qc = QuantumCircuit(qr, qr_ancilla)
+
+        asian_barrier_spread.build(qc, qr, qr_ancilla)
+        job = execute(qc, backend=BasicAer.get_backend('statevector_simulator'))
+
+        # evaluate resulting statevector
+        value = 0
+        for i, amplitude in enumerate(job.result().get_statevector()):
+            b = ('{0:0%sb}' % asian_barrier_spread.num_target_qubits).format(
+                i)[-asian_barrier_spread.num_target_qubits:]
+            prob = np.abs(amplitude)**2
+            if prob > 1e-4 and b[0] == '1':
+                value += prob
+                # all other states should have zero probability due to ancilla qubits
+                if i > 2**num_req_qubits:
+                    break
+
+        # map value to original range
+        mapped_value = asian_barrier_spread.value_to_estimation(
+            value) / (2**num_uncertainty_qubits - 1) * (high_ - low_)
+        expected = 0.83188
+        self.assertAlmostEqual(mapped_value, expected, places=5)
+
+
+@ddt
+class TestWeightedAdder(QiskitTestCase):
+    """Test the weighted adder circuit."""
+
+    def assertSummationIsCorrect(self, adder):
+        """Assert that ``adder`` correctly implements the summation w.r.t. it's set weights."""
+
+        circuit = QuantumCircuit(adder.num_qubits)
+        circuit.h(list(range(adder.num_state_qubits)))
+        circuit.append(adder.to_instruction(), list(range(adder.num_qubits)))
+
+        backend = BasicAer.get_backend('statevector_simulator')
+        statevector = execute(circuit, backend).result().get_statevector()
+
+        probabilities = defaultdict(float)
+        for i, statevector_amplitude in enumerate(statevector):
+            i = bin(i)[2:].zfill(circuit.num_qubits)[adder.num_ancilla_qubits:]
+            probabilities[i] += np.real(np.abs(statevector_amplitude) ** 2)
+
+        expectations = defaultdict(float)
+        for x in range(2**adder.num_state_qubits):
+            bits = np.array(list(bin(x)[2:].zfill(adder.num_state_qubits)), dtype=int)
+            summation = bits.dot(adder.weights[::-1])
+
+            entry = bin(summation)[2:].zfill(adder.num_sum_qubits) \
+                + bin(x)[2:].zfill(adder.num_state_qubits)
+            expectations[entry] = 1 / 2 ** adder.num_state_qubits
+
+        for state, probability in probabilities.items():
+            self.assertAlmostEqual(probability, expectations[state])
+
+    @data(
+        [0],
+        [1, 2, 1],
+        [4],
+    )
+    def test_summation(self, weights):
+        """Test the weighted adder on some examples."""
+        adder = WeightedAdder(len(weights), weights)
+        self.assertSummationIsCorrect(adder)
+
+    def test_mutability(self):
+        """Test the mutability of the weighted adder."""
+        adder = WeightedAdder()
+
+        with self.subTest(msg='missing number of state qubits'):
+            with self.assertRaises(AttributeError):
+                print(adder.draw())
```

Comment:
why draw the circuit as part of the testsuite?

Reply:
they are never drawn since the `.draw` method accesses `.data` which builds the circuit and then it raises. I just needed some use-case that calls `.data`. I can change it to 
```
_ = adder.data
```
but I thought the print resembles more the typical use-case we would have.